linera-service/src/cli/command.rs

```diff
@@ -194,6 +194,135 @@ use crate::util::{
+    #[serde(default)]
+    pub add: Vec<ValidatorSpec>,
+    #[serde(default)]
+    pub modify: Vec<ValidatorSpec>,
+    #[serde(default)]
+    pub remove: Vec<ValidatorPublicKey>,
+}
+
+/// Validator specification from JSON
+#[derive(Clone, Debug, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ValidatorSpec {
+    pub public_key: ValidatorPublicKey,
+    pub account_key: AccountPublicKey,
+    pub network_address: String,
+    pub votes: u64,
+}
+
+impl From<ValidatorSpec> for ValidatorToAdd {
+    fn from(spec: ValidatorSpec) -> Self {
+        ValidatorToAdd {
+            public_key: spec.public_key,
+            account_key: spec.account_key,
+            address: spec.network_address,
+            votes: spec.votes,
+        }
+    }
+}
+
+#[derive(Clone, clap::Subcommand)]
+pub enum ValidatorCommand {
+    /// Query a validator's status and compatibility
+    Query {
+        /// The validator's network address (e.g., grpcs://validator.net:443)
+        address: String,
+
+        /// The chain to query. If omitted, query the default chain of the wallet.
+        chain_id: Option<ChainId>,
+
+        /// The public key of the validator. If given, the signature will be checked.
+        #[arg(long)]
+        public_key: Option<ValidatorPublicKey>,
+    },
+
+    /// Query multiple validators from a JSON file
+    QueryBatch {
+        /// Path to JSON file containing validator list
+        file: PathBuf,
+
+        /// The chain to query. If omitted, query the default chain of the wallet.
+        #[arg(long)]
+        chain_id: Option<ChainId>,
+    },
+
+    /// List all validators for a chain
+    List {
+        /// The chain to query. If omitted, query the default chain of the wallet.
+        chain_id: Option<ChainId>,
+
+        /// Skip validators with less voting weight than this.
+        #[arg(long)]
+        min_votes: Option<u64>,
+    },
+
+    /// Add a new validator (admin only)
+    Add {
+        /// The public key of the validator.
+        #[arg(long)]
+        public_key: ValidatorPublicKey,
+
+        /// The public key of the account controlled by the validator.
+        #[arg(long)]
+        account_key: AccountPublicKey,
+
+        /// Network address
+        #[arg(long)]
+        address: String,
+
+        /// Voting power
+        #[arg(long, default_value = "1")]
+        votes: u64,
+
+        /// Skip the version and genesis config checks.
+        #[arg(long)]
+        skip_online_check: bool,
+    },
+
+    /// Remove a validator (admin only)
+    Remove {
+        /// The public key of the validator.
+        #[arg(long)]
+        public_key: ValidatorPublicKey,
+    },
+
+    /// Add, modify, and/or remove validators from a JSON file (admin only)
+    ///
+    /// JSON format supports two styles:
+    /// 1. Simple list: {"validators": [{...}]} - adds all validators
+    /// 2. Operations: {"add": [{...}], "modify": [{...}], "remove": ["pubkey"]}
```

Comment:
This is dangerous because the keys of the object are unordered so the commands may be executed in an unexpected order. I'd prefer JSON-lines:
```json
{ "add": … }
{ "modify": … }
{ "add": … }
…

Reply:
Alternatively, rather than introducing a new format here we can take a stream of lines of `update` commands in the same format as the individual commands. `--batch` can then be a flag to the normal update command to tell it to read a stream of argument lines.

Reply:
Why not a shell script? I haven't seen this approach anywhere before.